<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#API-Foursquare." data-toc-modified-id="API-Foursquare.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>API Foursquare.</a></span></li><li><span><a href="#Mi-empresa" data-toc-modified-id="Mi-empresa-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Mi empresa</a></span><ul class="toc-item"><li><span><a href="#Primera-llamada:-Aeropuerto" data-toc-modified-id="Primera-llamada:-Aeropuerto-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Primera llamada: Aeropuerto</a></span><ul class="toc-item"><li><span><a href="#El-aeropuerto-más-cercano-es-:" data-toc-modified-id="El-aeropuerto-más-cercano-es-:-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>El aeropuerto más cercano es :</a></span></li></ul></li><li><span><a href="#Segunda-llamada:-Nurseries" data-toc-modified-id="Segunda-llamada:-Nurseries-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Segunda llamada: Nurseries</a></span><ul class="toc-item"><li><span><a href="#¿Cuál-de-las-Nurseries-está-más-cerca-de-mi-empresa?" data-toc-modified-id="¿Cuál-de-las-Nurseries-está-más-cerca-de-mi-empresa?-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>¿Cuál de las Nurseries está más cerca de mi empresa?</a></span></li><li><span><a href="#La-Nursery-más-cercana-es-:" data-toc-modified-id="La-Nursery-más-cercana-es-:-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>La Nursery más cercana es :</a></span></li></ul></li><li><span><a href="#Tercera-llamada:-Nightclubs" data-toc-modified-id="Tercera-llamada:-Nightclubs-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Tercera llamada: Nightclubs</a></span><ul class="toc-item"><li><span><a href="#¿Dónde-está-el-Nightclub-más-cercano-a-mi-empresa?" data-toc-modified-id="¿Dónde-está-el-Nightclub-más-cercano-a-mi-empresa?-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>¿Dónde está el Nightclub más cercano a mi empresa?</a></span></li><li><span><a href="#El-nightclub-más-cercano-es-:" data-toc-modified-id="El-nightclub-más-cercano-es-:-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>El nightclub más cercano es :</a></span></li></ul></li><li><span><a href="#Cuarta-llamada:-Starbucks" data-toc-modified-id="Cuarta-llamada:-Starbucks-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Cuarta llamada: Starbucks</a></span><ul class="toc-item"><li><span><a href="#El-Starbucks-más-cercano-es-:" data-toc-modified-id="El-Starbucks-más-cercano-es-:-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>El Starbucks más cercano es :</a></span></li></ul></li><li><span><a href="#Quinta-llamada:-Restaurante-Vegano" data-toc-modified-id="Quinta-llamada:-Restaurante-Vegano-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Quinta llamada: Restaurante Vegano</a></span><ul class="toc-item"><li><span><a href="#¿Cuál-es-el-restaurante-vegano-más-cercano-a-mi-empresa?" data-toc-modified-id="¿Cuál-es-el-restaurante-vegano-más-cercano-a-mi-empresa?-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>¿Cuál es el restaurante vegano más cercano a mi empresa?</a></span></li><li><span><a href="#El-Restaurante-vegano-más-cercano-es-:" data-toc-modified-id="El-Restaurante-vegano-más-cercano-es-:-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>El Restaurante vegano más cercano es :</a></span></li></ul></li></ul></li></ul></div>

# API Foursquare.

Una vez seleccionada la oficina perfecta que cumple nuestras necesidades vamos a comprobar mediante la API de Foursquare los siguientes puntos;
- Los Account managers viajan mucho. Voy a busar un aeropuerto en un radio de 20kms.
- El 30% de los empleados de la compañía tiene al menos un niño. Voy a buscar guarderías en un radio de 5km.
- La edad de los trabajadores oscila entre 25 y 40 años. Voy a buscar discotecas en un radio de 10km.
- A los ejecutivos les gusta mucho la cadena de cafés "Starbucks". Voy a buscar Starbucks en un radio de 300m.
- La CEO de la compañía es vegana. Voy a buscar restaurantes veganos en un radio de 2km.

In [1]:
import requests
import json
from dotenv import load_dotenv
import os
import pandas as pd
from functools import reduce
import operator
import geopandas as gpd
from pymongo import MongoClient,GEOSPHERE
import shapely.geometry
import geopandas as gpd
from cartoframes.viz import Map, Layer, popup_element
import sys
sys.path.append("../")
import src.funcionesok as fun

In [2]:
company = "21 Mikonkatu, Helsinki FI 00100"

In [3]:
fun.geocode(company)

{'type': 'Point', 'coordinates': [60.17212, 24.94519]}

In [ ]:
fun.geo_api(company)

In [4]:
load_dotenv()

True

In [5]:
tok1 = os.getenv("tok1")
tok2 = os.getenv("tok2")

In [6]:
suomi = pd.read_csv('../data/dataOK.csv', encoding = 'ISO - 8859-1')
suomi

city country_code   latitude  longitude   name category_code  \
0  Helsinki          FIN  60.171821  24.945167  Apaja   games_video   

   number_of_employees  founded_year  
0                   44          2001

In [7]:
#Busco los códigos en la API en los que voy a basar las llamadas;

"""
airport = "4bf58dd8d48988d1ed931735"

nurseries = "4f4533814b9074f6e4fb0107"

nightlife sport = "4d4b7105d754a06376d81259"

coffee_shop = "4bf58dd8d48988d1e0931735"

vegan_restaurant = "4bf58dd8d48988d1d3941735"

"""

'\nairport = "4bf58dd8d48988d1ed931735"\n\nnurseries = "4f4533814b9074f6e4fb0107"\n\nnightlife sport = "4d4b7105d754a06376d81259"\n\ncoffee_shop = "4bf58dd8d48988d1e0931735"\n\nvegan_restaurant = "4bf58dd8d48988d1d3941735"\n\n'

#  Mi empresa
Teniendo en cuenta mi empresa hago llamadas a la API.

## Primera llamada: Aeropuerto

In [8]:
donde = "21 Mikonkatu, Helsinki FI 00100"
def geocode(address):
    """
    Saca las coordenadas de una dirección que le des.
    """
    data = requests.get(f"https://geocode.xyz/{address}?json=1").json()
    try:
        return {
            "type":"Point",
            "coordinates":[float(data["longt"]),float(data["latt"])]}
    except:
        return data

In [9]:
emp1 = geocode(donde)
emp1

{'type': 'Point', 'coordinates': [24.94519, 60.17212]}

In [10]:
url1= 'https://api.foursquare.com/v2/venues/explore'

In [11]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"{emp1.get('coordinates')[1]},{emp1.get('coordinates')[0]}",
              "query": "Airport",
              "limit": 1,
              "radius": 20000,
             }

In [12]:
resp = requests.get(url=url1, params=params)
airdata = json.loads(resp.text)


In [13]:
airdata.keys()

dict_keys(['meta', 'response'])

In [14]:
decoding_data = airdata.get("response")

In [15]:
decoded = decoding_data.get("groups")[0]

In [16]:
airp = decoded.get("items")

In [17]:
mapa_nombre = ["venue","name"]
m_latitud = ["venue","location","lat"]
m_longitud = ["venue","location","lng"]

In [18]:
def getFromDict(diccionario,mapa):
    return reduce (operator.getitem,mapa,diccionario)

In [19]:
print(getFromDict(airp[0],mapa_nombre))

Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)


In [20]:
print(getFromDict(airp[0],m_latitud))

60.31912855684579


In [21]:
print(getFromDict(airp[0],m_longitud))

24.968297481536865


In [22]:
aeropuerto = []
for dic in airp:
    ai = {}
    ai["name"] = getFromDict(dic,mapa_nombre)
    ai["latitud"] = getFromDict(dic,m_latitud)
    ai["longitud"] = getFromDict(dic,m_longitud)
    aeropuerto.append(ai)

In [23]:
aeropuerto[0]

{'name': 'Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)',
 'latitud': 60.31912855684579,
 'longitud': 24.968297481536865}

In [24]:
dataframefinal = pd.DataFrame(aeropuerto)
dataframefinal.head()

name    latitud   longitud
0  Helsinki Airport (HEL) (Helsinki-Vantaan lento...  60.319129  24.968297

In [25]:
gdf = gpd.GeoDataFrame(dataframefinal, geometry = gpd.points_from_xy(dataframefinal.longitud,dataframefinal.latitud))
gdf.head()

name    latitud   longitud  \
0  Helsinki Airport (HEL) (Helsinki-Vantaan lento...  60.319129  24.968297   

                    geometry  
0  POINT (24.96830 60.31913)

In [26]:
#Creo un mapa con el punto selecccionado

In [27]:
Map(Layer(gdf, popup_hover = [popup_element("name","Airport")]))

### El aeropuerto más cercano es : 
    Helsinki Airport (HEL) (Helsinki-Vantaan lentoasema)
    latitud: 60.319129
    longitud: 24.968297

## Segunda llamada: Nurseries

In [28]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"{emp1.get('coordinates')[1]},{emp1.get('coordinates')[0]}",
              "query": "Nursery School",
              "limit": 10,
              "radius": 5000
             }

In [29]:
resp = requests.get(url=url1, params=params)
nurdata = json.loads(resp.text)

In [30]:
nurdata.keys()

dict_keys(['meta', 'response'])

In [31]:
decoding_data = nurdata.get("response")

In [32]:
decoded = decoding_data.get("groups")[0]

In [33]:
nur = decoded.get("items")


In [34]:
mapan_nombre = ["venue","name"]
mn_latitud = ["venue","location","lat"]
mn_longitud = ["venue","location","lng"]

In [35]:
def getFromDict(diccionario,mapa):
    return reduce (operator.getitem,mapa,diccionario)

In [36]:
print(getFromDict(nur[0],mapan_nombre))
print(getFromDict(nur[0],mn_latitud))
print(getFromDict(nur[0],mn_longitud))

Daghemmet B. E. Von Schantz
60.158121
24.936681


In [37]:
nurseries = []
for dic in nur:
    nr = {}
    nr["name"] = getFromDict(dic,mapa_nombre)
    nr["latitud"] = getFromDict(dic,m_latitud)
    nr["longitud"] = getFromDict(dic,m_longitud)
    nurseries.append(nr)

In [38]:
nurseries[0]

{'name': 'Daghemmet B. E. Von Schantz',
 'latitud': 60.158121,
 'longitud': 24.936681}

In [39]:
df_nur = pd.DataFrame(nurseries)
df_nur.head()

name    latitud   longitud
0     Daghemmet B. E. Von Schantz  60.158121  24.936681
1  lauttasaari english playschool  60.158857  24.877429
2             Päiväkoti Kesäheinä  60.202455  24.951274
3               Päiväkoti Marjala  60.209475  24.964812

In [40]:
import geopandas as gpd

In [41]:
gdf2 = gpd.GeoDataFrame(df_nur, geometry = gpd.points_from_xy(df_nur.longitud,df_nur.latitud))
gdf2.head()

name    latitud   longitud  \
0     Daghemmet B. E. Von Schantz  60.158121  24.936681   
1  lauttasaari english playschool  60.158857  24.877429   
2             Päiväkoti Kesäheinä  60.202455  24.951274   
3               Päiväkoti Marjala  60.209475  24.964812   

                    geometry  
0  POINT (24.93668 60.15812)  
1  POINT (24.87743 60.15886)  
2  POINT (24.95127 60.20246)  
3  POINT (24.96481 60.20948)

In [42]:
Map(Layer(gdf2, popup_hover = [popup_element("name","Nursery School")]))

### ¿Cuál de las Nurseries está más cerca de mi empresa?

In [43]:
gdf2['geometry']=gdf2['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf2

/usr/local/Caskroom/miniconda/base/envs/ironhack/lib/python3.8/site-packages/geopandas/geodataframe.py:852: UserWarning: Geometry column does not contain geometry.
  warnings.warn("Geometry column does not contain geometry.")


name    latitud   longitud  \
0     Daghemmet B. E. Von Schantz  60.158121  24.936681   
1  lauttasaari english playschool  60.158857  24.877429   
2             Päiväkoti Kesäheinä  60.202455  24.951274   
3               Päiväkoti Marjala  60.209475  24.964812   

                                            geometry  
0  {'type': 'Point', 'coordinates': (24.936681, 6...  
1  {'type': 'Point', 'coordinates': (24.877429136...  
2  {'type': 'Point', 'coordinates': (24.951274170...  
3  {'type': 'Point', 'coordinates': (24.964811971...

In [44]:
client = MongoClient()
db = client.ironhack
collection = db.nurseries
collection.create_index ([("geometry", GEOSPHERE)])

'geometry_2dsphere'

In [45]:
data = gdf2.to_dict(orient='records')
collection.insert_many(data)

In [46]:
data

[{'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': (24.936681, 60.158121)},
  '_id': ObjectId('602164a3bda0fd886bdf4fe1')},
 {'name': 'lauttasaari english playschool',
  'latitud': 60.158857316461074,
  'longitud': 24.877429136253458,
  'geometry': {'type': 'Point',
   'coordinates': (24.877429136253458, 60.158857316461074)},
  '_id': ObjectId('602164a3bda0fd886bdf4fe2')},
 {'name': 'Päiväkoti Kesäheinä',
  'latitud': 60.20245514603548,
  'longitud': 24.95127417000266,
  'geometry': {'type': 'Point',
   'coordinates': (24.95127417000266, 60.20245514603548)},
  '_id': ObjectId('602164a3bda0fd886bdf4fe3')},
 {'name': 'Päiväkoti Marjala',
  'latitud': 60.20947535503745,
  'longitud': 24.964811971036774,
  'geometry': {'type': 'Point',
   'coordinates': (24.964811971036774, 60.20947535503745)},
  '_id': ObjectId('602164a3bda0fd886bdf4fe4')}]

In [47]:
cercano2 = db.nurseries.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano2)

[{'_id': ObjectId('60212ef3235f9908d38c87d0'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('602164a3bda0fd886bdf4fe1'),
  'name': 'Daghemmet B. E. Von Schantz',
  'latitud': 60.158121,
  'longitud': 24.936681,
  'geometry': {'type': 'Point', 'coordinates': [24.936681, 60.158121]}},
 {'_id': ObjectId('602164a3bda0fd886bdf4fe4'),
  'name': 'Päiväkoti Marjala',
  'latitud': 60.20947535503745,
  'longitud': 24.964811971036774,
  'geometry': {'type': 'Point',
   'coordinates': [24.964811971036774, 60.20947535503745]}},
 {'_id': ObjectId('60212ef3235f9908d38c87d3'),
  'name': 'Päiväkoti Marjala',
  'latitud': 60.20947535503745,
  'longitud': 24.964811971036774,
  'geometry': {'type': 'Point',
   'coordinates': [24.964811971036774, 60.20947535503745]}},
 {'_id': ObjectId('602164a3bda0fd886bdf4fe3'),
  'name': 'Päiväkoti Kesäheinä',
  'latitud': 60.2024551460

### La Nursery más cercana es : 
    Daghemmet B. E. Von Schantz
    latitud': 60.158121
    longitud': 24.936681

## Tercera llamada: Nightclubs

In [48]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"{emp1.get('coordinates')[1]},{emp1.get('coordinates')[0]}",
              "query": "Nightlife Spot",
              "limit": 25,
              "radius": 7500 
         }
resp = requests.get(url=url1, params=params)
nidata = json.loads(resp.text)

In [49]:
nidata.keys()

dict_keys(['meta', 'response'])

In [50]:
decoding_data = nidata.get("response")

In [51]:
decoded = decoding_data.get("groups")[0]


In [52]:
nic = decoded.get("items")


In [53]:
mapani_nombre = ["venue","name"]
mapani_latitud = ["venue","location","lat"]
mapani_longitud = ["venue","location","lng"]

In [54]:
print(getFromDict(nic[0],mapa_nombre))
print(getFromDict(nic[0],m_latitud))
print(getFromDict(nic[0],m_longitud))

Liberty or Death
60.16469982956184
24.944248795509335


In [55]:
nightclubs = []
for dic in nic:
    nc = {}
    nc["name"] = getFromDict(dic,mapa_nombre)
    nc["latitud"] = getFromDict(dic,m_latitud)
    nc["longitud"] = getFromDict(dic,m_longitud)
    nightclubs.append(nc)

In [56]:
nightclubs[0]

{'name': 'Liberty or Death',
 'latitud': 60.16469982956184,
 'longitud': 24.944248795509335}

In [57]:
df_nc = pd.DataFrame(nightclubs)
df_nc.head()

name    latitud   longitud
0          Liberty or Death  60.164700  24.944249
1                 KirjaClub  60.176467  24.956222
2                     Apéro  60.164211  24.929481
3  Olutravintola Pikkulintu  60.165893  24.936865
4   Plaza Wine & Craft Beer  60.169200  24.934983

In [58]:
gdf3 = gpd.GeoDataFrame(df_nc, geometry = gpd.points_from_xy(df_nc.longitud,df_nc.latitud))
gdf3.head()

name    latitud   longitud                   geometry
0          Liberty or Death  60.164700  24.944249  POINT (24.94425 60.16470)
1                 KirjaClub  60.176467  24.956222  POINT (24.95622 60.17647)
2                     Apéro  60.164211  24.929481  POINT (24.92948 60.16421)
3  Olutravintola Pikkulintu  60.165893  24.936865  POINT (24.93687 60.16589)
4   Plaza Wine & Craft Beer  60.169200  24.934983  POINT (24.93498 60.16920)

In [59]:
Map(Layer(gdf3, popup_hover = [popup_element("name","Night Club")]))

### ¿Dónde está el Nightclub más cercano a mi empresa?

In [60]:
gdf3['geometry']=gdf3['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf3

/usr/local/Caskroom/miniconda/base/envs/ironhack/lib/python3.8/site-packages/geopandas/geodataframe.py:852: UserWarning: Geometry column does not contain geometry.
  warnings.warn("Geometry column does not contain geometry.")


name    latitud   longitud  \
0           Liberty or Death  60.164700  24.944249   
1                  KirjaClub  60.176467  24.956222   
2                      Apéro  60.164211  24.929481   
3   Olutravintola Pikkulintu  60.165893  24.936865   
4    Plaza Wine & Craft Beer  60.169200  24.934983   
5                     Kaisla  60.171878  24.947378   
6                      Runar  60.166330  24.949431   
7               Sori Taproom  60.171963  24.947435   
8                        Bro  60.177036  24.951549   
9                 Bar Llamas  60.163469  24.942379   
10                Bar Favela  60.172056  24.920077   
11           Pub Peräkammari  60.167797  24.937183   
12          Punavuoren Ahven  60.161082  24.937536   
13                Café Bar 9  60.164911  24.941953   
14        Viinibaari Vin-Vin  60.167683  24.939196   
15                 Cafe Talo  60.180181  24.950721   
16      Navy Jerry's Rum Bar  60.169658  24.926351   
17                 Bar Haven  60.166545  24.951354   
18                  Kämp Bar  60.167816  24.947222   
19           Chihuahua Julep  60.165137  24.943945   
20              One Pint Pub  60.162014  24.920067   
21             Room Bulevard  60.163807  24.936436   
22         Viinibaari Apotek  60.166580  24.932240   
23     The Bull and the Firm  60.188338  24.964497   
24          Pinehill Terrace  60.183647  24.928906   

                                             geometry  
0   {'type': 'Point', 'coordinates': (24.944248795...  
1   {'type': 'Point', 'coordinates': (24.956222434...  
2   {'type': 'Point', 'coordinates': (24.929480552...  
3   {'type': 'Point', 'coordinates': (24.936865199...  
4   {'type': 'Point', 'coordinates': (24.934983, 6...  
5   {'type': 'Point', 'coordinates': (24.947377580...  
6   {'type': 'Point', 'coordinates': (24.949430865...  
7   {'type': 'Point', 'coordinates': (24.947435376...  
8   {'type': 'Point', 'coordinates': (24.951548759...  
9   {'type': 'Point', 'coordinates': (24.942379295...  
10  {'type': 'Point', 'coordinates': (24.920077373...  
11  {'type': 'Point', 'coordinates': (24.937183114...  
12  {'type': 'Point', 'coordinates': (24.937535539...  
13  {'type': 'Point', 'coordinates': (24.941953489...  
14  {'type': 'Point', 'coordinates': (24.939196, 6...  
15  {'type': 'Point', 'coordinates': (24.950721057...  
16  {'type': 'Point', 'coordinates': (24.926351387...  
17  {'type': 'Point', 'coordinates': (24.951353670...  
18  {'type': 'Point', 'coordinates': (24.947221502...  
19  {'type': 'Point', 'coordinates': (24.943945, 6...  
20  {'type': 'Point', 'coordinates': (24.920066560...  
21  {'type': 'Point', 'coordinates': (24.936436213...  
22  {'type': 'Point', 'coordinates': (24.93224, 60...  
23  {'type': 'Point', 'coordinates': (24.964497, 6...  
24  {'type': 'Point', 'coordinates': (24.928905588...

In [61]:
db = client.ironhack
collection = db.nightclubs
collection.create_index ([("geometry", GEOSPHERE)])

'geometry_2dsphere'

In [62]:
data = gdf3.to_dict(orient='records')
collection.insert_many(data)

In [63]:
data[:5]

[{'name': 'Liberty or Death',
  'latitud': 60.16469982956184,
  'longitud': 24.944248795509335,
  'geometry': {'type': 'Point',
   'coordinates': (24.944248795509335, 60.16469982956184)},
  '_id': ObjectId('602164b3bda0fd886bdf4fe5')},
 {'name': 'KirjaClub',
  'latitud': 60.176467430909206,
  'longitud': 24.956222434547158,
  'geometry': {'type': 'Point',
   'coordinates': (24.956222434547158, 60.176467430909206)},
  '_id': ObjectId('602164b3bda0fd886bdf4fe6')},
 {'name': 'Apéro',
  'latitud': 60.16421142755892,
  'longitud': 24.92948055267334,
  'geometry': {'type': 'Point',
   'coordinates': (24.92948055267334, 60.16421142755892)},
  '_id': ObjectId('602164b3bda0fd886bdf4fe7')},
 {'name': 'Olutravintola Pikkulintu',
  'latitud': 60.1658932577839,
  'longitud': 24.93686519981724,
  'geometry': {'type': 'Point',
   'coordinates': (24.93686519981724, 60.1658932577839)},
  '_id': ObjectId('602164b3bda0fd886bdf4fe8')},
 {'name': 'Plaza Wine & Craft Beer',
  'latitud': 60.1692,
  'longitud

In [64]:
cercano3 = db.nightclubs.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano3)[:5]

[{'_id': ObjectId('60212f89235f9908d38c87e5'),
  'name': 'Bar Haven',
  'latitud': 60.16654471941996,
  'longitud': 24.9513536704468,
  'geometry': {'type': 'Point',
   'coordinates': [24.9513536704468, 60.16654471941996]}},
 {'_id': ObjectId('60213075235f9908d38c87fe'),
  'name': 'Bar Haven',
  'latitud': 60.16654471941996,
  'longitud': 24.9513536704468,
  'geometry': {'type': 'Point',
   'coordinates': [24.9513536704468, 60.16654471941996]}},
 {'_id': ObjectId('602164b3bda0fd886bdf4ff6'),
  'name': 'Bar Haven',
  'latitud': 60.16654471941996,
  'longitud': 24.9513536704468,
  'geometry': {'type': 'Point',
   'coordinates': [24.9513536704468, 60.16654471941996]}},
 {'_id': ObjectId('60212f89235f9908d38c87da'),
  'name': 'Runar',
  'latitud': 60.16632954758476,
  'longitud': 24.949430865771262,
  'geometry': {'type': 'Point',
   'coordinates': [24.949430865771262, 60.16632954758476]}},
 {'_id': ObjectId('60213075235f9908d38c87f3'),
  'name': 'Runar',
  'latitud': 60.16632954758476,
  

### El nightclub más cercano es : 
    Bar Haven
    latitud': 60.16654471941996
    longitud': 24.9513536704468

## Cuarta llamada: Starbucks

In [65]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"{emp1.get('coordinates')[1]},{emp1.get('coordinates')[0]}",
              "query": "Starbucks",
              "limit": 10,
              "radius": 500
         }
resp = requests.get(url=url1, params=params)
strdata = json.loads(resp.text)


In [66]:
strdata.keys()

dict_keys(['meta', 'response'])

In [67]:
decoding_data = strdata.get("response")

In [68]:
decoded = decoding_data.get("groups")[0]

In [69]:
star = decoded.get("items")

In [70]:
mapastar_nombre = ["venue","name"]
mapastar_latitud = ["venue","location","lat"]
mapastar_longitud = ["venue","location","lng"]

In [71]:
print(getFromDict(star[0],mapastar_nombre))
print(getFromDict(star[0],mapastar_latitud))
print(getFromDict(star[0],mapastar_longitud))

Starbucks
60.167736932209785
24.94336882366579


In [72]:
starbucks = []
for dic in star:
    st = {}
    st["name"] = getFromDict(dic,mapa_nombre)
    st["latitud"] = getFromDict(dic,m_latitud)
    st["longitud"] = getFromDict(dic,m_longitud)
    starbucks.append(st)

In [73]:
starbucks[0]

{'name': 'Starbucks',
 'latitud': 60.167736932209785,
 'longitud': 24.94336882366579}

In [74]:
#Al comprobar los starbucks, me doy cuenta que existe sólo uno en Helsinki y me muestra otras categorías. 
df_st = pd.DataFrame(starbucks)
df_st.head()

name    latitud   longitud
0               Starbucks  60.167737  24.943369
1         Robert's Coffee  60.169206  24.941611
2              Café Aalto  60.167825  24.943540
3  Robert's Coffee Netcup  60.168703  24.940922
4         Robert's Coffee  60.171241  24.941166

In [75]:
gdf4 = gpd.GeoDataFrame(df_st, geometry = gpd.points_from_xy(df_st.longitud,df_st.latitud))
gdf4.head()

name    latitud   longitud                   geometry
0               Starbucks  60.167737  24.943369  POINT (24.94337 60.16774)
1         Robert's Coffee  60.169206  24.941611  POINT (24.94161 60.16921)
2              Café Aalto  60.167825  24.943540  POINT (24.94354 60.16783)
3  Robert's Coffee Netcup  60.168703  24.940922  POINT (24.94092 60.16870)
4         Robert's Coffee  60.171241  24.941166  POINT (24.94117 60.17124)

In [76]:
Map(Layer(gdf4, popup_hover = [popup_element("name","Starbucks")]))

### El Starbucks más cercano es : 
    
    Starbucks
    latitud: 60.167737
    longitud: 24.943369

## Quinta llamada: Restaurante Vegano

In [77]:
params = {"client_id" : tok1,
              "client_secret" : tok2,
              "v": "20180323",
              "ll": f"{emp1.get('coordinates')[1]},{emp1.get('coordinates')[0]}",
              "query": "Vegan Restaurant",
              "limit": 20,
              "radius": 2000
         }
resp = requests.get(url=url1, params=params)
vegdata = json.loads(resp.text)


In [78]:
decoding_data = vegdata.get("response")

In [79]:
decoded = decoding_data.get("groups")[0]

In [80]:
vegg = decoded.get("items")

In [81]:
mapavegg_nombre = ["venue","name"]
mapavegg_latitud = ["venue","location","lat"]
mapavegg_longitud = ["venue","location","lng"]

In [82]:
print(getFromDict(vegg[0],mapavegg_nombre))
print(getFromDict(vegg[0],mapavegg_latitud))
print(getFromDict(vegg[0],mapavegg_longitud))

Zucchini
60.16616033742283
24.949550935785922


In [83]:
vegano = []
for dic in vegg:
    vg = {}
    vg["name"] = getFromDict(dic,mapavegg_nombre)
    vg["latitud"] = getFromDict(dic,mapavegg_latitud)
    vg["longitud"] = getFromDict(dic,mapavegg_longitud)
    vegano.append(vg)

In [84]:
vegano[0]

{'name': 'Zucchini',
 'latitud': 60.16616033742283,
 'longitud': 24.949550935785922}

In [85]:
df_vg = pd.DataFrame(vegano)
df_vg.head()

name    latitud   longitud
0       Zucchini  60.166160  24.949551
1  Roots Kitchen  60.179254  24.951582
2          Qulma  60.171590  24.956639
3    Yes Yes Yes  60.164187  24.943998
4         Levant  60.164131  24.936531

In [86]:
gdf5 = gpd.GeoDataFrame(df_vg, geometry = gpd.points_from_xy(df_vg.longitud,df_vg.latitud))
gdf5.head()

name    latitud   longitud                   geometry
0       Zucchini  60.166160  24.949551  POINT (24.94955 60.16616)
1  Roots Kitchen  60.179254  24.951582  POINT (24.95158 60.17925)
2          Qulma  60.171590  24.956639  POINT (24.95664 60.17159)
3    Yes Yes Yes  60.164187  24.943998  POINT (24.94400 60.16419)
4         Levant  60.164131  24.936531  POINT (24.93653 60.16413)

In [87]:
Map(Layer(gdf5, popup_hover = [popup_element("name","Venga Restaurant")]))

### ¿Cuál es el restaurante vegano más cercano a mi empresa?

In [88]:
gdf5['geometry']=gdf5['geometry'].apply(lambda x:shapely.geometry.mapping(x))
gdf5

/usr/local/Caskroom/miniconda/base/envs/ironhack/lib/python3.8/site-packages/geopandas/geodataframe.py:852: UserWarning: Geometry column does not contain geometry.
  warnings.warn("Geometry column does not contain geometry.")


name    latitud   longitud  \
0               Zucchini  60.166160  24.949551   
1          Roots Kitchen  60.179254  24.951582   
2                  Qulma  60.171590  24.956639   
3            Yes Yes Yes  60.164187  24.943998   
4                 Levant  60.164131  24.936531   
5                   Magu  60.163830  24.946299   
6              Silvoplee  60.181398  24.949064   
7               Sen Chay  60.177595  24.950797   
8                 Fafa's  60.169105  24.933750   
9                  OmNam  60.168093  24.935987   
10       Classic & Vegan  60.179440  24.950385   
11      Eat Your Greens!  60.169102  24.948195   
12            Loving Hut  60.182881  24.948376   
13                Levant  60.188394  24.954584   
14             Just Vege  60.188386  24.957649   
15          OHMYGOODNESS  60.188234  24.955744   
16                Pontus  60.158792  24.946282   
17      The Red Koi Thai  60.164367  24.932645   
18  Luomupuoti Satumarja  60.179978  24.950927   
19               Bär Bar  60.164984  24.936220   

                                             geometry  
0   {'type': 'Point', 'coordinates': (24.949550935...  
1   {'type': 'Point', 'coordinates': (24.951581663...  
2   {'type': 'Point', 'coordinates': (24.956639065...  
3   {'type': 'Point', 'coordinates': (24.943998, 6...  
4   {'type': 'Point', 'coordinates': (24.936530690...  
5   {'type': 'Point', 'coordinates': (24.946299, 6...  
6   {'type': 'Point', 'coordinates': (24.949064133...  
7   {'type': 'Point', 'coordinates': (24.950796621...  
8   {'type': 'Point', 'coordinates': (24.933750430...  
9   {'type': 'Point', 'coordinates': (24.935987093...  
10  {'type': 'Point', 'coordinates': (24.950385, 6...  
11  {'type': 'Point', 'coordinates': (24.948195268...  
12  {'type': 'Point', 'coordinates': (24.948376411...  
13  {'type': 'Point', 'coordinates': (24.954584387...  
14  {'type': 'Point', 'coordinates': (24.957649483...  
15  {'type': 'Point', 'coordinates': (24.955744, 6...  
16  {'type': 'Point', 'coordinates': (24.946281948...  
17  {'type': 'Point', 'coordinates': (24.932644517...  
18  {'type': 'Point', 'coordinates': (24.950927495...  
19  {'type': 'Point', 'coordinates': (24.93622, 60...

In [89]:
db = client.ironhack
collection = db.veganos
collection.create_index ([("geometry", GEOSPHERE)])

'geometry_2dsphere'

In [90]:
data = gdf5.to_dict(orient='records')
collection.insert_many(data)

In [91]:
data[:5]

[{'name': 'Zucchini',
  'latitud': 60.16616033742283,
  'longitud': 24.949550935785922,
  'geometry': {'type': 'Point',
   'coordinates': (24.949550935785922, 60.16616033742283)},
  '_id': ObjectId('602164e5bda0fd886bdf4ffe')},
 {'name': 'Roots Kitchen',
  'latitud': 60.17925390601578,
  'longitud': 24.951581663171815,
  'geometry': {'type': 'Point',
   'coordinates': (24.951581663171815, 60.17925390601578)},
  '_id': ObjectId('602164e5bda0fd886bdf4fff')},
 {'name': 'Qulma',
  'latitud': 60.17158999776495,
  'longitud': 24.95663906516804,
  'geometry': {'type': 'Point',
   'coordinates': (24.95663906516804, 60.17158999776495)},
  '_id': ObjectId('602164e5bda0fd886bdf5000')},
 {'name': 'Yes Yes Yes',
  'latitud': 60.164187,
  'longitud': 24.943998,
  'geometry': {'type': 'Point', 'coordinates': (24.943998, 60.164187)},
  '_id': ObjectId('602164e5bda0fd886bdf5001')},
 {'name': 'Levant',
  'latitud': 60.164130694350206,
  'longitud': 24.93653069038648,
  'geometry': {'type': 'Point',
   '

In [92]:
cercano5 = db.veganos.find({"geometry":{"$near":{"type":"Point", "coordinates":[60.171821,24.945167]}}})
list(cercano5)[:5]

[{'_id': ObjectId('602137dc235f9908d38c8816'),
  'name': 'Pontus',
  'latitud': 60.1587920757416,
  'longitud': 24.94628194859813,
  'geometry': {'type': 'Point',
   'coordinates': [24.94628194859813, 60.1587920757416]}},
 {'_id': ObjectId('602164e5bda0fd886bdf500e'),
  'name': 'Pontus',
  'latitud': 60.1587920757416,
  'longitud': 24.94628194859813,
  'geometry': {'type': 'Point',
   'coordinates': [24.94628194859813, 60.1587920757416]}},
 {'_id': ObjectId('602137dc235f9908d38c880b'),
  'name': 'Magu',
  'latitud': 60.16383,
  'longitud': 24.946299,
  'geometry': {'type': 'Point', 'coordinates': [24.946299, 60.16383]}},
 {'_id': ObjectId('602164e5bda0fd886bdf5003'),
  'name': 'Magu',
  'latitud': 60.16383,
  'longitud': 24.946299,
  'geometry': {'type': 'Point', 'coordinates': [24.946299, 60.16383]}},
 {'_id': ObjectId('602137dc235f9908d38c8806'),
  'name': 'Zucchini',
  'latitud': 60.16616033742283,
  'longitud': 24.949550935785922,
  'geometry': {'type': 'Point',
   'coordinates': [

### El Restaurante vegano más cercano es : 
    
    Pontus
    latitud: 60.1587920757416
    longitud: 24.94628194859813